### PR TITLE
Add specific product_carriers in webservice products

### LINKED ISSUE #18277: Id_Reference In the Carrier list.
#### Is your feature request related to a problem? Yes

I added in the webservice the possibility to set specific carriers for products.
The table product_carrier need Id_Refence of the carrier but this id is not show in the carrier list (in backoffice).
Id_Reference is the true id of the carrier and Id_carrier is most like the "version" !
If you make any change in a carrier, a new carrier is recorded (to keep an history).
The version before is set to deleted = yes.
The Id_Reference is the link between the old and new version.

#### Describe the solution you'd like

I think just adding ID_Reference should in the carrier list will help users to setup their ERP.

#### Alternatives you've considered

I tryed to replace Id_Reference by Id_carrier in webservice but it's not a good idea because for any changed word in the carrier you must change ERP parameters.
In fact, the curent version is set to "deleted" and a new carrier, with a new id !, is recorded. 

#### Additional context

see getWsCarriers and setWsCarriers in Product class.

### PR DESCRIPTION
**Pull request by jcvoyeux**
Add product_carriers in webservices

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add specific product_carriers in webservice products
| Type?         | improvement
| Category?     | WS 
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #18277.
| How to test?  | https:/yoursite/api/products/aidproduct

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18343)

### CODE REVIEW

**Comment by PierreRambaud** [comment]
Hi and thanks for your contribution.
Before beginning the review, you need to add a proper title, good description, fill the table correctly, fix the CI (using php-cs-fixer) ;)

Kind regards

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
See comment

**Comment by jcvoyeux** [comment]
Hello,

I see there was a conflict.
I resolved it but could you verify everything is OK ?
Thank you

Take care of you 😉

________________________________
De : GoT <notifications@github.com>
Envoyé : vendredi 27 mars 2020 14:31
À : PrestaShop/PrestaShop <PrestaShop@noreply.github.com>
Cc : jcvoyeux <jcvoyeux@hotmail.com>; Mention <mention@noreply.github.com>
Objet : Re: [PrestaShop/PrestaShop] Update Product.php (#18343)


@PierreRambaud requested changes on this pull request.

See comment

File: classes/Product.php [hunk 1]
```diff
@@ -537,8 +537,17 @@ class ProductCore extends ObjectModel
...
             ],
+            'carriers' => [
+        				'resource' => 'carrier',
+                'getter' => 'getWsCarriers',
+                'setter' => 'setWsCarriers',	
+                'fields' => [
+                    'id' => ['required' => true],
+                    'id_shop' => ['required' => true],
+                ],
+            ],
```
Comment (Progi1984) [inline classes/Product.php:548]:
```suggestion
            'carriers' => [
        		'resource' => 'carrier',
                'getter' => 'getWsCarriers',
                'setter' => 'setWsCarriers',
                'fields' => [
                    'id' => ['required' => true],
                    'id_shop' => ['required' => true],
                ],
            ],
```

File: classes/Product.php [hunk 2]
```diff
@@ -537,8 +537,17 @@ class ProductCore extends ObjectModel
...
+                'getter' => 'getWsCarriers',
+                'setter' => 'setWsCarriers',	
+                'fields' => [
+                    'id' => ['required' => true],
+                    'id_shop' => ['required' => true],
+                ],
+            ],            
         ],
-    ];
+    ];
```
Comment (Progi1984) [inline classes/Product.php:550]:
```suggestion
    ];
```

File: classes/Product.php [hunk 3]
```diff
@@ -6944,6 +6953,54 @@ public function modifierWsLinkRewrite()
 
         return true;
     }
+    
+	/*
+	* For webservices (use id_reference not id_carrier)
+	* Get the liste of specific carriers for a product
```
Comment (Progi1984) [inline classes/Product.php:6959]:
```suggestion
	* Get the list of specific carriers for a product
```

File: classes/Product.php [hunk 4]
```diff
@@ -6944,6 +6953,54 @@ public function modifierWsLinkRewrite()
...
+	* Don't know if I must limit to the default shop ?? (AND pc.id_shop = ' . (int) $this->id_shop_default)
```
Comment (Progi1984) [inline classes/Product.php:6960]:
```suggestion
	* Don't know if I must limit to the default shop ?? (AND pc.id_shop = ' . (int) $this->id_shop_default)
```

Think MultiShop ;)

File: classes/Product.php [hunk 5]
```diff
@@ -6944,6 +6953,54 @@ public function modifierWsLinkRewrite()
...
+			
+        return $carriers;
+    }
+	
+	/*
+	* For webservices (use id_reference not id_carrier)
+	* Delete old carrieres
+	* For each carrierr 
+	*	If found and not setted 'deleted' -> insert it
+	* Don't know if I must limit to the default shop ?? (pc.id_shop = ' . (int) $this->id_shop_default)
```
Comment (Progi1984) [inline classes/Product.php:6979]: Think MultiShop

File: classes/Product.php [hunk 6]
```diff
@@ -6944,6 +6953,54 @@ public function modifierWsLinkRewrite()
...
+            LEFT JOIN `' . _DB_PREFIX_ . 'carrier` c ON (pc.id_carrier_reference = c.id_reference) AND (c.deleted = 0)
+            WHERE (pc.id_product = ' . (int) $this->id . ')'
+            );
+			
+        return $carriers;
+    }
+	
+	/*
+	* For webservices (use id_reference not id_carrier)
+	* Delete old carrieres
```
Comment (Progi1984) [inline classes/Product.php:6976]:
```suggestion
	* Delete old carriers
```

File: classes/Product.php [hunk 7]
```diff
@@ -6944,6 +6953,54 @@ public function modifierWsLinkRewrite()
...
+	* For each carrierr
```
Comment (Progi1984) [inline classes/Product.php:6977]:
```suggestion
	* For each carrier
```

File: classes/Product.php [hunk 8]
```diff
@@ -6944,6 +6953,54 @@ public function modifierWsLinkRewrite()
...
+	* For webservices (use id_reference not id_carrier)
+	* Delete old carrieres
+	* For each carrierr 
+	*	If found and not setted 'deleted' -> insert it
+	* Don't know if I must limit to the default shop ?? (pc.id_shop = ' . (int) $this->id_shop_default)
+	*/
+	public function setWsCarriers($carriers)
+	{
+		$answ = true;
+		Db::getInstance()->execute('DELETE FROM `' . _DB_PREFIX_ . 'product_carrier`
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
		Db::getInstance()->execute(
		    'DELETE FROM `' . _DB_PREFIX_ . 'product_carrier`
```

**Comment by jcvoyeux** [comment]
Must I do something ?
Message "Process completed with exit code 8" ??

**Comment by Progi1984** [comment]
@jcvoyeux You can try in local with this command line : `./vendor/bin/php-cs-fixer fix --dry-run --diff`

**Comment by colinegin** [comment]
Hello @jcvoyeux ,

Are you able to fix your PR according to @Progi1984 's comments ?

**Comment by PierreRambaud** [comment]
Hi @jcvoyeux ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**